File: 0016-3sum-closest/0016-3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums.sort()
        closest = nums[0] + nums[1] + nums[2]

        for i in range(len(nums)):
            if i > 0 and nums[i] == nums[i - 1]: continue # Skip duplicates
            start = i + 1
            end = len(nums) - 1
            while start < end:
                total = nums[i] + nums[start] + nums[end]
                if abs(target - total) < abs(target - closest):
                    closest = total
                if total < target:
                    start += 1
                elif total > target:
                    end -= 1
                else:
                    return total

        return closest

# Synced seamlessly with LeetHub Pro
# Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
# Get it here: https://chromewebstore.google.com/detail/bcilpkkbokcopmabingnndookdogmbna

File: 0016-3sum-closest/0016-3sum-closest.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var threeSumClosest = function(nums, target) {
    nums.sort((a, b) => a - b)

    nearest = nums[0] + nums[1] + nums[2]
    for(let i = 0; i<nums.length; i++){
        if (i > 0 && nums[i] == nums[i - 1]) {continue}

        let start = i + 1
        let end = nums.length - 1

        while (start < end){
            total = nums[i] + nums[start] + nums[end]

            if (Math.abs(target - total) < Math.abs(target - nearest)){
                nearest = total
            }
            if (total === target){
                return total
            } else if (total < target){
                start += 1
            } else {
                end -= 1
            }
        }
    }
    return nearest
};

